hazelcast/src/main/java/com/hazelcast/sql/impl/client/ReadOptimizedLruCache.java

```diff
@@ -0,0 +1,151 @@
+ * of the eldest items in two passes.
+ * <p>
+ * The cleanup process isn't synchronized to guarantee that the capacity is not
+ * exceeded. The cache is available during the cleanup for reads and writes. If
+ * there's a large number of writes by many threads, the one thread doing the
+ * cleanup might not be quick enough and there's no upper bound on the actual
+ * size of the cache. This is done to optimize the happy path when the keys fit
+ * into the cache.
+ */
+public class ReadOptimizedLruCache<K, V> {
+
+    // package-visible for tests
+    final ConcurrentMap<K, ValueAndTimestamp<V>> cache;
+
+    private final AtomicBoolean cleanupLock = new AtomicBoolean();
+    private final int capacity;
+    private final int cleanupThreshold;
+
+    /**
+     * @param cleanupThreshold The size at which the cache will clean up oldest
+     *     entries in batch. `cleanupThreshold - capacity` entries will be removed.
+     */
+    public ReadOptimizedLruCache(int capacity, int cleanupThreshold) {
+        if (capacity <= 0) {
+            throw new IllegalArgumentException("capacity <= 0");
+        }
+        if (cleanupThreshold <= capacity) {
+            throw new IllegalArgumentException("cleanupThreshold <= capacity");
+        }
+
+        this.capacity = capacity;
+        this.cleanupThreshold = cleanupThreshold;
+
+        cache = new ConcurrentHashMap<>(cleanupThreshold);
+    }
+
+    public V getOrDefault(K key, V defaultValue) {
+        final V existingValue = get(key);
+        return existingValue != null ? existingValue : defaultValue;
+    }
+
+    public V get(K key) {
+        ValueAndTimestamp<V> valueFromCache = cache.get(key);
+        if (valueFromCache == null) {
+            return null;
+        }
+        valueFromCache.touch();
+        return valueFromCache.value;
+    }
+
+    public void put(K key, V value) {
+        if (value == null) {
+            throw new IllegalArgumentException("Null values are disallowed");
+        }
+
+        ValueAndTimestamp<V> oldValue = cache.put(key, new ValueAndTimestamp<>(value));
+        if (oldValue == null && cache.size() > cleanupThreshold) {
+            doCleanup();
+        }
+    }
+
+    public void remove(K key) {
+        cache.remove(key);
+    }
+
+    private void doCleanup() {
+        // if no thread is cleaning up, we'll do it
+        if (!cleanupLock.compareAndSet(false, true)) {
+            return;
+        }
+
+        try {
+            int entriesToRemove = cache.size() - capacity;
+            if (entriesToRemove <= 0) {
+                // this can happen if the cache is concurrently modified
+                return;
+            }
+            PriorityQueue<Long> oldestTimestamps =
+                    new PriorityQueue<>(entriesToRemove + 1, Comparator.<Long>naturalOrder().reversed());
+
+            // 1st pass
+            for (ValueAndTimestamp<V> valueAndTimestamp : cache.values()) {
```

Comment:
```suggestion
            for (ValueAndTimestamp<V> valueAndTimestamp : cache.values()) {
                oldestTimestamps.add(valueAndTimestamp.timestamp);
            }
            
            for (int i= 1; i< entriesToRemove; i++){
                oldestTimestamps.poll();
            }
```

If you test with different size of cache and threshold value, you will see it fails with current implementation. I'm suggesting to remove oldest items after queue is completed.

Reply:
Okay, this is my silliness. The used priority queue here is **max-heap** not min-heap. So, it resolves everything. I should see that `reversed` call at the begging.